File: board-room/src/main/java/net/stzups/board/server/websocket/protocol/server/messages/ServerMessageUpdateCanvas.java
package net.stzups.board.server.websocket.protocol.server.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.board.data.objects.canvas.Canvas;
import net.stzups.board.server.websocket.Client;
import net.stzups.board.server.websocket.protocol.server.ServerMessage;
import net.stzups.board.server.websocket.protocol.server.ServerMessageType;

import java.util.Map;

public class ServerMessageUpdateCanvas extends ServerMessage {
    private Map<Client, Canvas> canvasMap;

    public ServerMessageUpdateCanvas(Map<Client, Canvas> canvasMap) {
        super(ServerMessageType.UPDATE_CANVAS);
        this.canvasMap = canvasMap;
    }

    public void serialize(ByteBuf byteBuf) {
        super.serialize(byteBuf);
        byteBuf.writeShort((short) canvasMap.size());
        for (Map.Entry<Client, Canvas> entry : canvasMap.entrySet()) {
            byteBuf.writeShort(entry.getKey().getId());
            entry.getValue().serialize(byteBuf);
        }
    }
}

File: frontend/html/scripts/protocol/server/messages/ServerMessageUpdateCanvas.js
import ServerMessage from "../ServerMessage.js";
import ServerMessageType from "../ServerMessageType.js";
import {Canvas} from "../../../canvas/Canvas.js";

export default class ServerMessageUpdateCanvas extends ServerMessage {
    constructor(reader) {
        super(ServerMessageType.UPDATE_CANVAS);
        this.canvasMap = new Map();
        let length = reader.readUint16();
        for (let i = 0; i < length; i++) {
            this.canvasMap.put(reader.readInt16(), new Canvas(reader));
        }
    }
}